mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/persistence/KeyValueBasedPersistenceProvider.java

```diff
@@ -0,0 +1,718 @@
+import io.mantisrx.server.master.domain.DataFormatAdapter;
+import io.mantisrx.server.master.domain.JobClusterDefinitionImpl.CompletedJob;
+import io.mantisrx.server.master.resourcecluster.ClusterID;
+import io.mantisrx.server.master.resourcecluster.TaskExecutorID;
+import io.mantisrx.server.master.resourcecluster.TaskExecutorRegistration;
+import io.mantisrx.server.master.store.InvalidJobException;
+import io.mantisrx.server.master.store.KeyValueStore;
+import io.mantisrx.server.master.store.MantisJobMetadataWritable;
+import io.mantisrx.server.master.store.MantisStageMetadata;
+import io.mantisrx.server.master.store.MantisStageMetadataWritable;
+import io.mantisrx.server.master.store.MantisWorkerMetadataWritable;
+import io.mantisrx.server.master.store.NamedJob;
+import io.mantisrx.shaded.com.fasterxml.jackson.core.JsonProcessingException;
+import io.mantisrx.shaded.com.fasterxml.jackson.core.type.TypeReference;
+import io.mantisrx.shaded.com.fasterxml.jackson.databind.DeserializationFeature;
+import io.mantisrx.shaded.com.fasterxml.jackson.databind.ObjectMapper;
+import io.mantisrx.shaded.com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
+import io.mantisrx.shaded.com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
+import io.mantisrx.shaded.com.google.common.collect.ImmutableList;
+import io.mantisrx.shaded.com.google.common.collect.Lists;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import rx.Observable;
+
+
+/**
+ * Provide a key-value aware implementation of IMantisStorageProvider.
+ * This has all the table-models and mantis specific logic for how
+ * mantis master should store job, cluster and related metadata
+ * assuming the underlying storage provides key based lookups.
+ *
+ * The instance of this class needs an actual key-value storage
+ * implementation (implements KeyValueStorageProvider) to function.
+ *
+ * Look at {@code io.mantisrx.server.master.store.KeyValueStorageProvider}
+ * to see the features needed from the storage.
+ *   Effectively, an apache-cassandra like storage with primary key made
+ *   up of partition key and composite key and a way to iterate over all
+ *   partition keys in the table.
+ */
+public class KeyValueBasedPersistenceProvider implements IMantisPersistenceProvider {
```

Comment:
Should this class be abstract?

Reply:
no this is a concrete implementation. 
Another way of writing this code is to make this class abstract and move the interface methods from KeyValueStore here but then that'd lead to more bloating as all persistence use-cases would have to end up here.